coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+> In this example it might make more sense if the owner name was `readonly` and the owner had both a setter and a getter.
+
+### [Private Variables](#private-variables)
+Prefix private variables with `_`, e.g. `_name`, `_age`, `_height`.
+
+### [Naming Guidelines](#naming-guidelines)
+Naming convention is important for others reading your code and contributing. A lot of emphasis should be put on correctly naming a function or a variable.
+```typescript
+// DON'T MAKE A CLASS LIKE THIS!
+class Kek {
+  private readonly m: number;
+  private bleurg: number = 0;
+  constructor(m: number){
+    this.m = m;
+  }
+  
+  add(n: number): boolean {
+    if(this.bleurg + n < this.m) {
+      this.bleurg =+ n;
+      return true;
+    }
+    return false;
+    
+  }
+}
+```
+Excuse the hyperbole example, but even the `add` function which has a descriptive name does do additional work which someone unfamiliar with the code will not catch. The more logic that gets entered into a function, the more likely it is that the name of the function no longer describes the code.
+
+```typescript
+// DON'T MAKE A CLASS LIKE THIS!
+class Wallet {
+  private _amount: number;
+  private _loan: number = 0;
+  
+  constructor(amount: number) {
+    this._amount = amount;
+  }
+
+  pay(price: number): boolean {
+    if(this._amount - price > 0) {
+      this._amount -= price;
+      return true;
+    } else if(this._loan < 5000 && this._amount !== 0) {
+      const loanAmount = price - this._amount;
+      this._amount = 0;
+      loan += loanAmount;
+      return true;
+    } else if(this._loan < 5000 && this._amount === 0) {
+      loan += price;
+      return true;
+    }
+    return false;
+  }
+}
+```
+> **Note:** In addition to the naming convention this price doesn't check if the price number is > 0 to verify that they aren't adding funcs into the wallet buy "paying" for negative values.
+
+The above example illustrates in a way that the function name is no longer discriptive of what the code does. It also starts to blend into a section covered later [Single Responsibility Pattern](#single-responsibility-pattern), but first let's cover the naming convention. A way to make the function more discriptive would be:
+```typescript
+...
+payFromWalletOrAttemptToLendMoneyToBuy(price: number): boolean {
+  ...
+}
+...
+```
+Now the function name mirrors to an extend the behavior of the function, but in a way the function is still a bit wrong. We return a boolean to see if the purchase was accepted, so a better name might have been.
+```typescript
+...
+successfullyPaidFromWalletOrLentMoneyToBuy(price: number): boolean {
+  ...
+}
+...
+```
+Now in addition to describing what happens inside the function it also matches with the return type of the function.
+
+# Arrays
+
+### [Array Instantiation](#array-instantiation)
+For array instantiation you should use the literal syntax instead of `new` constructor.
+```typescript
+// const items: number[] = new Array(); Don't do this.
```

Comment:
Makes perfect sense to use `new Array(...)`, e.g. when you want to specify size up front

Reply:
No, actually that is bad. Saying new Array(5); actually doesn't create an array with the size 5, it just creates an array with length 5. This can actually trick you.

First of all you can no longer use length to see if an array is empty.
I would completely agree if there was some merit into doing this, but the way Arrays are implemented in js seems to not strengthen this.

https://stackoverflow.com/questions/18947892/creating-range-in-javascript-strange-syntax

Basically, if you do it this way, you can't find out if an array is empty, since lenght != 0 this way.